![UKDS Logo](./images/UKDS_Logos_Col_Grey_300dpi.png)

# Social Network Analysis: Fundamental Concepts

Welcome to the <a href="https://ukdataservice.ac.uk/" target=_blank>UK Data Service</a> training series on *New Forms of Data for Social Science Research*. This series guides you through some of the most common and valuable new sources of data available for social science research: data collected from websites, social media platorms, text data, conducting simulations (agent based modelling), to name a few. To help you get to grips with these new forms of data, we provide webinars, interactive notebooks containing live programming code, reading lists and more.

* To access training materials for the entire series: <a href="https://github.com/UKDataServiceOpen/new-forms-of-data" target=_blank>[Training Materials]</a>

* To keep up to date with upcoming and past training events: <a href="https://ukdataservice.ac.uk/news-and-events/events" target=_blank>[Events]</a>

* To get in contact with feedback, ideas or to seek assistance: <a href="https://ukdataservice.ac.uk/help.aspx" target=_blank>[Help]</a>

<a href="https://www.research.manchester.ac.uk/portal/julia.kasmire.html" target=_blank>Dr Julia Kasmire</a> and <a href="https://www.research.manchester.ac.uk/portal/diarmuid.mcdonnell.html" target=_blank>Dr Diarmuid McDonnell</a> <br />
UK Data Service  <br />
University of Manchester <br />
September 2020

<h1>Table of Contents<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Introduction" data-toc-modified-id="Introduction-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Introduction</a></span><ul class="toc-item"><li><span><a href="#Aims" data-toc-modified-id="Aims-1.1"><span class="toc-item-num">1.1&nbsp;&nbsp;</span>Aims</a></span></li><li><span><a href="#Lesson-details" data-toc-modified-id="Lesson-details-1.2"><span class="toc-item-num">1.2&nbsp;&nbsp;</span>Lesson details</a></span></li></ul></li><li><span><a href="#Guide-to-using-this-resource" data-toc-modified-id="Guide-to-using-this-resource-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Guide to using this resource</a></span><ul class="toc-item"><li><span><a href="#Interaction" data-toc-modified-id="Interaction-2.1"><span class="toc-item-num">2.1&nbsp;&nbsp;</span>Interaction</a></span></li><li><span><a href="#Learn-more" data-toc-modified-id="Learn-more-2.2"><span class="toc-item-num">2.2&nbsp;&nbsp;</span>Learn more</a></span></li></ul></li><li><span><a href="#Overview" data-toc-modified-id="Overview-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>Overview</a></span><ul class="toc-item"><li><span><a href="#What-is-Social-Network-Analysis-(SNA)?" data-toc-modified-id="What-is-Social-Network-Analysis-(SNA)?-3.1"><span class="toc-item-num">3.1&nbsp;&nbsp;</span>What is Social Network Analysis (SNA)?</a></span></li><li><span><a href="#What-are-the-principles-behind-SNA?" data-toc-modified-id="What-are-the-principles-behind-SNA?-3.2"><span class="toc-item-num">3.2&nbsp;&nbsp;</span>What are the principles behind SNA?</a></span></li><li><span><a href="#Why-should-you-consider-SNA-for-your-research?" data-toc-modified-id="Why-should-you-consider-SNA-for-your-research?-3.3"><span class="toc-item-num">3.3&nbsp;&nbsp;</span>Why should you consider SNA for your research?</a></span></li><li><span><a href="#When-should-you-use-SNA?" data-toc-modified-id="When-should-you-use-SNA?-3.4"><span class="toc-item-num">3.4&nbsp;&nbsp;</span>When should you use SNA?</a></span></li><li><span><a href="#What-does-SNA-involve?" data-toc-modified-id="What-does-SNA-involve?-3.5"><span class="toc-item-num">3.5&nbsp;&nbsp;</span>What does SNA involve?</a></span></li><li><span><a href="#How-do-you-implement-SNA-in-your-research?" data-toc-modified-id="How-do-you-implement-SNA-in-your-research?-3.6"><span class="toc-item-num">3.6&nbsp;&nbsp;</span>How do you implement SNA in your research?</a></span></li></ul></li><li><span><a href="#Key-Concepts" data-toc-modified-id="Key-Concepts-4"><span class="toc-item-num">4&nbsp;&nbsp;</span>Key Concepts</a></span><ul class="toc-item"><li><span><a href="#Entities" data-toc-modified-id="Entities-4.1"><span class="toc-item-num">4.1&nbsp;&nbsp;</span>Entities</a></span></li><li><span><a href="#Connections" data-toc-modified-id="Connections-4.2"><span class="toc-item-num">4.2&nbsp;&nbsp;</span>Connections</a></span></li><li><span><a href="#Components" data-toc-modified-id="Components-4.3"><span class="toc-item-num">4.3&nbsp;&nbsp;</span>Components</a></span></li><li><span><a href="#Networks" data-toc-modified-id="Networks-4.4"><span class="toc-item-num">4.4&nbsp;&nbsp;</span>Networks</a></span></li></ul></li><li><span><a href="#Representing-Networks-as-Graphs-and-Matrices" data-toc-modified-id="Representing-Networks-as-Graphs-and-Matrices-5"><span class="toc-item-num">5&nbsp;&nbsp;</span>Representing Networks as Graphs and Matrices</a></span><ul class="toc-item"><li><span><a href="#Graphs" data-toc-modified-id="Graphs-5.1"><span class="toc-item-num">5.1&nbsp;&nbsp;</span>Graphs</a></span></li><li><span><a href="#Matrices" data-toc-modified-id="Matrices-5.2"><span class="toc-item-num">5.2&nbsp;&nbsp;</span>Matrices</a></span></li></ul></li><li><span><a href="#A-Simple-Analysis" data-toc-modified-id="A-Simple-Analysis-6"><span class="toc-item-num">6&nbsp;&nbsp;</span>A Simple Analysis</a></span></li><li><span><a href="#Conclusion" data-toc-modified-id="Conclusion-7"><span class="toc-item-num">7&nbsp;&nbsp;</span>Conclusion</a></span></li><li><span><a href="#Bibliography" data-toc-modified-id="Bibliography-8"><span class="toc-item-num">8&nbsp;&nbsp;</span>Bibliography</a></span></li><li><span><a href="#Further-reading-and-resources" data-toc-modified-id="Further-reading-and-resources-9"><span class="toc-item-num">9&nbsp;&nbsp;</span>Further reading and resources</a></span></li><li><span><a href="#Appendices" data-toc-modified-id="Appendices-10"><span class="toc-item-num">10&nbsp;&nbsp;</span>Appendices</a></span></li></ul></div>

## Introduction

Vast swathes of our social interactions and personal behaviours are now conducted online and/or captured digitally. Thus, computational methods for collecting, cleaning and analysing data are an increasingly important component of a social scientist’s toolkit.

In this training series we cover some of the essential knowledge and skills needed to engage in **Social Network Analysis (SNA)**, a methodological approach that provides concepts, tools and techniques for uncovering and understanding social structures, relations and networks of assocation. We focus on the three major stages of SNA:
1. Understanding fundamental concepts and terms [Focus of this notebook].
2. Collecting and cleaning social network data from various sources.
3. Performing basic and intermediate analyses of social network data. 

By the end of these lessons you should be confident in your understanding of key SNA concepts and terms, proficient in the handling and cleaning of social network data, and able to apply a range of analytical techniques to derive substantive insight about social structures and relations. In addition, you will gain fluency in the use of the Python programming language for SNA and other computational social science tasks.

### Aims

This lesson - **Social Network Analysis: Fundamental Concepts** - has two aims:
1. Define and examine fundamental concepts and terms underpinning SNA.
2. Cultivate your computational skills through coding examples. For example, there are a number of opportunities for you to interactively explore key concepts, as well as examine data stored in a network format.

### Lesson details

* **Level**: Introductory, for individuals with no prior knowledge or experience of social network analysis.
* **Duration**: 30-60 minutes.
* **Pre-requisites**: None.
* **Audience**: Researchers and analysts from any disciplinary background interested in employing network analysis for social science research purposes.
* **Programming language**: Python.
* **Learning outcomes**:
	1. Understand fundamental concepts and terms associated with SNA.
	2. Understand how social network data are structured and represented.

## Guide to using this resource

This learning resource was built using <a href="https://jupyter.org/" target=_blank>Jupyter Notebook</a>, an open-source software application that allows you to mix code, results and narrative in a single document. As <a href="https://jupyter4edu.github.io/jupyter-edu-book/" target=_blank>Barba et al. (2019)</a> espouse:
> In a world where every subject matter can have a data-supported treatment, where computational devices are omnipresent and pervasive, the union of natural language and computation creates compelling communication and learning opportunities.

If you are familiar with Jupyter notebooks then skip ahead to the main content (*What is Social Network Analysis?*). Otherwise, the following is a quick guide to navigating and interacting with the notebook.

### Interaction

**You only need to execute the code that is contained in sections which are marked by `In []`.**

To execute a cell, click or double-click the cell and press the `Run` button on the top toolbar (you can also use the keyboard shortcut Shift + Enter).

Try it for yourself:

In [ ]:
print("Enter your name and press enter:")
name = input()
print("\r")
print("Hello {}, enjoy learning more about social network analysis!".format(name))

### Learn more

Jupyter notebooks provide rich, flexible features for conducting and documenting your data analysis workflow. To learn more about additional notebook features, we recommend working through some of the <a href="https://github.com/darribas/gds19/blob/master/content/labs/lab_00.ipynb" target=_blank>materials</a> provided by Dani Arribas-Bel at the University of Liverpool. 

## Overview

Social networks are commonplace in modern life, mainly due to the influence of social media platforms like Twitter, Facebook, Instagram and many others, though there are also a multitude of networks arising from offline interactions also e.g., business/academic networking, community groups organised around causes/neighbourhood projects etc (Scott, 2017).

Terminology

Unit of Analysis

### What is Social Network Analysis (SNA)?

Social network analysis (SNA) is a methodological and conceptual toolbox for the measurement, systematic description, and analysis of patterns in relational structures in the social world (Schneider, 2008). A relation is a distinctive type of connection or tie between two entities (Wasserman & Faust, 1994). For example, a married couple share a spousal relation, a brother and sister share a sibling relation, co-workers share a collegial relation etc. Relations are the building blocks of networks, and thus SNA is concerned with and most appropriate for analyses of data capturing relations between units of analysis (Scott, 2017).

[*Key point that needs to be said somewhere - does it go in the principles section?*]"That is, the network analyst is always interested in how the individual is embedded within a structure and how the structure emerges from the micro-relations between individual parts." (7 of PDF)

It originated from *network theory* - a branch of *graph theory* - which seeks to generate measurable representations of patterns of relationships between entities in an abstract or actual space (Owen-Smith, 2017). As a result of these origins, SNA is a highly technical and mathematical approach to sociological analysis, replete with an intimidating vocabularly of terms and concepts not otherwise encountered with other forms of social science research (Scott, 2017). In addition, the data underpinning SNA have a distinctive structure.

This lesson will provide a clear, concise guide to these concepts and terms, as well as introduce you to social network data. 

### What are the principles behind SNA?

Borgatti et al. (2002) establish five key principles of SNA:
1. *SNA focuses on relations (connections) between actors.* Actors and their relations are seen as interdependent rather than independent units.
2. *The relations between actors are the most meaningful focus of analysis.* Your data may allow you to perform other types of analyses - e.g., how does income vary across individuals and is it associated with variation in subjective wellbeing? - but the focus of SNA is on examining and understanding how actors are connected e.g., to what extent are individuals connected and are these patterns associated with income and subjective wellbeing? Put another way, the central concern of analytical concern is the relation, not the individual.
3. *The structural and/or relational features of these actors constitute the analytically relevant characteristics of them.* To quote Freeman (2006, p. #):
>...these patterns [interactions between people] are important features of the lives of the individuals who display them.
4. *Relational ties between these actors are the channels for the flow of both material and non-material resources*. In essence, the connections are important and offer opportunities for sharing of valuable resources.
5. *The complete web of actors, their positions and their linkages - the network structure - provides opportunities for (and constraints upon) action.* This is one of the most important points to remember: networks are not only constructed from the relations between actors, they in turn influence the behaviour, opportunities, contraints and outcomes of said actors. We'll reflect on some of the mechanisms through which networks affect outcomes at the micro level in a later lesson [LINK]().

It is this focus on relational rather than attributional aspects of the units of analysis that makes SNA distinct as a methodology (Caiani, 2014).

### Why should you consider SNA for your research?

From an analytical perspective, SNA can be employed for a variety of valuable purposes (Caiani, 2014; Owen-Smith, 2017):
* The social phenomenon of interest is a network i.e., it is the unit of analysis in your research project. For example, a researcher may be interested in analysing the London Underground rail network (like in this [study](https://doi.org/10.1016/j.jtrangeo.2017.11.018)).
* The features and properties of a network can be important explanatory factors ('right-hand side' variables) for understanding other social phenomena. For example, in a review of the impact of social networks on health outcomes, Smith and Christakis (2008, p. 420) conclude that:
>  illness, disability, health behaviors, health care use, and death in one person are associated with similar outcomes in numerous others to whom that person is tied, and there can be a nonbiological transmission of illness.

### When should you use SNA?

When you are dealing with **relational data** i.e., data capturing relationships and connections between units of analysis. This is in contrast to **attributional data**, which captures the attributes of your units of analysis.

[Give Examples]

### What does SNA involve?

Employing SNA in your research typically involves the following activities (Scott, 2017):
* Identifying and visualising patterns of relations between units of analysis.
* Examining structural properties/characteristics of these relations.
* Analysing implications of these relations on outcomes experienced by units of analysis.

As a result of its focus on the relational characteristics of the units of analysis, SNA requires distinctive data structures, methods of analysis and data visualisation techniques (Caiani, 2014). 

### How do you implement SNA in your research?

There are a number of key steps in conducting social network analysis (Hanneman & Riddle, 2005; Owen-Smith, 2017):
1. Pose a carefully articulated research question that requires understanding and/or analysis of a network.
2. Decide which units of analysis and types of relations to analyse i.e., who is connected and which relationships matter? 
3. Collect or select a data set that provides relational data on your units of analysis. This data set can include attributional data also: for example, Twitter data can provide information on which accounts follow each other (relational), as well as details about the accounts themselves (attributional).
4. Summarise the network and its key features using appropriate measures e.g., network size, density, cohesion etc.

## Key Concepts

[*Need to use plenty of visual and tabular (matrix) examples throughout this section*]

[*At this point we want to switch to the language of nodes and ties, and give a reason up front for doing so*]

A network has two key components (Owen-Smith, 2017):
1. The entities that are (or can be) connected in a network.
2. The connections that exist (or could exist) between the entities.

For example, a family tree is a network containing individuals (entities) that are related through some type of familial tie (connection). Therefore a network is an aggregation or collection of these entities and their connections.

### Entities

The entities that are connected in a network are known as **nodes**, a term derived from network theory. Nodes can be individuals, organisations, countries, animals, events, computers, train stations etc. We could also refer to entities as *actors* or *agents* - which are terms specific to SNA -, or *vertices* - which is a term used in geometry -, or *points*, which is used in graph theory. For consistency, we will use the term **node**.

Two nodes that are or could be connected are called a **dyad**, while three nodes that are or could be connected are called a **triad**.

We'll cover this in more detail in a later lesson but for now it is worth posing the following questions about nodes:
* Who are they and how many of them are there in a network?
* What connections/ties exists between them?
* What positions do they occupy in the network? For example, do they broker connections between other nodes?

### Connections

[*Talk about direct and indirect ties, strong and weak ties*]

Connections or relations between entities are known as **ties**. Other terms include *edges* (network theory), *lines* (graph theory), or *links*. There are a multitude of different types of ties: for example, a family tree contains ties between brothers and sisters, nieces and uncles, children and parents etc. Two individuals can also be related in multiple ways: for example, a pair of colleagues may also be good friends, part of the same sports club, and have attended the same university. As stated previously, it is therefore crucial to clearly define which ties you are interested in measuring, and acknowledging that your data are most likely a sample of **all** possible ties that exist between your nodes (Hanneman & Riddle, 2005).

There are many measurement scales we can use to measure ties (Hanneman & Riddle, 2005):
* Nominal or Binary: presence or absence of a relation
* Multinomial: type of relation e.g., friend, colleague etc - usually transformed into binary.
* Ordinal (grouped): likert scale of relation strength or frequency e.g., on a scale of 1-5 (1 = weak, 5 = strong), classify the relationship you have with each family member.
* Ordinal (full-rank): rank nodes in order of relation e.g., best friend, second best friend.
* Interval: numerical scores assigned to relations e.g., relations between charities and funders, valued by how much funding is provided. It is possible to switch from interval to binary by specifying a threshold or cut point.

However many methods and measures of analysis focus on ties measured on a *binary* or *interval* scale, and thus we focus on these throughout the rest of this and future lessons.

We can define a ties as possessing two dimensions (Scott, 2017):
1. Numeration / Strength:
    * Binary: a tie exists between two entities
    * Valued: a tie can be assigned a value representing greater / lesser importance, strength, prominence etc. (This corresponds to the interval measurement scale)
2. Directionality:
    * Directed: the tie flows from one entity to another (and is potentially reciprocated); put another way, the relation has a *source node* and *target node*. For example, an individual (source) donates money to a charity (target).
    * Undirected: the tie does not originate from or terminate at a particular node e.g., if John is married to Jane, Jane must be married to John. Therefore an undirected tie *symmetric* or *reciprical* by default. It is also possible to treat directed ties as undirected: for example, I am sharing this lesson with you, therefore a connection exists between us.
    
We can combine these two dimensions to produce the following types of ties (give examples):
Arcs
Edges

### Components

Also known as a subgraph, a **component** is a specific subgroup of nodes and ties contained within the overall network. Technically, a component is a set of nodes where every node is maximally connected: that is, each node is connected to every other node, either directly or indirectly. In addition, there are no connections with nodes outside the component (Scott, 2017).

It can be hard to make sense of a component, so let's visualise one below. [*Draw example of ten nodes, seven of which are contained within a node*]

### Networks

[*What is a network?*] "It is often the case that the structure of a network depends on the attributes of the actors embedded in it."


Types of networks:
Whole = totality of connections between a set of nodes.
Ego-centric = defines a focal node (ego) and measures  its connections with other nodes (alters); we are usually interested in the connections between alters also.

Networks tend to be multi-modal i.e., nested within other networks (Hanneman & Riddle, 2005). For example, individuals are nested within organisations, which are nested within industries, which are nested within countries etc. Individuals may have ties to people in a different organisation, while the organisation may also have ties with other organisations. Therefore using our understanding of nodes (entities) and ties (connections), we can identify different network modalities:
* One-mode network contains one type of node (e.g., organisations) - this is also known as a *unipartite* network.
* Two-mode network contains two types of node (e.g., organisations and their employees) - this is also known as a *bipartite* or *affiliate* network.

A two-mode network can be conceptualised and analysed as two one-mode networks (projections) - this process is known as *inducing*. However, as Owen-Smith (2017, p. 223) warns: 
> Care must be taken when inducing one-mode network projections from two-mode network data because not all affiliations provide equally compelling evidence of actual social relationships.

For example, if we were convert our two-mode network of organisations and employees to a one-mode network of employees, it is likely to be very sparsely connected because most employees only work for one company.

A point to remember: network structure is not always the result of deliberate action on the part of the entities i.e., their is no vision or strategy for the network, it simply arises through the many small decisions of the entities.

## Representing Networks as Graphs and Matrices

Networks can be represented using two main, complementary methods:
1. Graphs
2. Matrices

### Graphs

Many people are most familiar with the visual representation of networks, known as **graphs** (network theory) or **sociograms** (SNA). A graph is a set of lines connecting points, and graph theory is a "*body of mathematical axioms and formulae that describe the properties of the patterns formed by the lines."* (Scott, 2017: #). A basic example is below: this graph represents an network containing 5 nodes and 10 ties: the ties are binary, undirected and of the same type (often referred to as a **simple graph**).

In [1]:
## INSERT CODE HERE FOR DIFFERENT TYPES OF GRAPHS

[*Show an example of a digraph etc*]

In a network graph (Hanneman & Riddle, 2005):
* Nodes are represented as circles
* Ties are represented as lines (with arrow heads if the tie is directed)
* Colours, shapes and sizes can be used to differentiate nodes by their attributes or network characteristics.
* Colours, shapes and sizes can also be used to differentiate relations by their type or amount.

For example, you could use thicker lines to represent stronger ties; colour coding to distinguish nodes by some attribute (European vs African country); or different line styles - dashed vs solid - to distnguish different types of ties (e.g., spousal or collegial).

Graphs are often a useful way of revealing network structure and properties, though visualisation is not a sufficient (or even) necessary method of analysing a network (Hanneman & Riddle, 2005). It is best practice to use visualisations in tandem with numerical summaries, descriptions and analyses. We'll revist this point in our lesson on analysing social networks.

### Matrices

A matrix is an arrangement of elements into rows (i) and columns (j). Matrices will be very familiar to anyone who has worked with data stored in a spreadsheet: each row represents an observation, each column a variable, and each cell a value for that variable. For example, here is a data set containing basic details for charities registered in Manchester:

In [61]:
import pandas as pd
import numpy as np # MOVE THESE TO A PRELIMINARY SECTION

data = pd.read_csv("./data/manchester-charities-2020-07-03.csv", index_col = False)
data.sample(5)

regno                                   Name  Latest income  \
131   506421        SLADE LANE NEIGHBOURHOOD CENTRE          33254   
8     222631                 THE EDWARD MAYES TRUST              0   
463  1100516   ISLAMIC SOCIAL AND EDUCATIONAL TRUST          31182   
324  1059115  GREATER MANCHESTER SPORTS PARTNERSHIP        1902651   
725  1143639       THE PORTICO LIBRARY AND NEWSROOM         196477   

     Latest expenditure                                            Address  \
131               32396  SLADE LANE NEIGHBOURHOOD CENTRE, 642 STOCKPORT...   
8                     0  STAFF OFFICE EDWARD MAYES TRUST, MAYES GARDENS...   
463               20897  ISLAMIC SOCIAL AND EDUCATION TRUST, St. Johns ...   
324             1900757  Greater Sport, National Squash Centre, Rowsley...   
725               58967  The Portico Library & Gallery, 57 Mosley Stree...   

    Postcode  
131  M13 0RZ  
8     M4 7FN  
463  M13 0WU  
324  M11 3FF  
725   M2 3FF

For instance, matrix X 

* Matrix X = i (rows) x j (columns)
* Transposed matrix X' = h x g
* One-mode network XX' = (g x h) * (h x g) = (g x g)
* One-mode network X'X = (h x g) * (g x h) = (h x h)

A useful heuristic is to think of the rows as the *source* of the tie and the columns as the *target* of the tie - this is particularly relevant when directed network data are stored in matrix format.

[*Move this entire section to the Simple Analysis and replace with a synthetic, simple analysis]

#### Relational data

We begin with data we collected about the current trustees (board members) of our group of manchester charities - see below.

In [94]:
data = pd.read_csv("./data/manchester-trustees-2020-08-13.csv", index_col = False)
data.head(12)

trustee_id          trustee_name  trustee_total_tships    regno
0         5873  Rabbi Abraham Hassan                     3  1071809
1         5873  Rabbi Abraham Hassan                     3  1095687
2         5873  Rabbi Abraham Hassan                     3  1013846
3        14229        David Neuwirth                     9  1123674
4        14229        David Neuwirth                     9  1166641
5        14229        David Neuwirth                     9  1083461
6        14229        David Neuwirth                     9  1109132
7        14229        David Neuwirth                     9  1084316
8        14229        David Neuwirth                     9  1136917
9        14229        David Neuwirth                     9  1183303
10       14229        David Neuwirth                     9  1111457
11       14229        David Neuwirth                     9  1092751

As you can see, the first individual is a trustee of three charities, the second a trustee of nine etc. Because we have the unique id (`regno`) of the charity a trustee is connected to, this data set contains *relational information* on how charities are connected to each other by the presence of a common individual. For instance, three of the charities - 1071809, 1095687, 1013846 - are all linked through the trustee Rabbi Abraham Hassan. 

Our task is to convert this data set into a case-by-case (adjacency) matrix. That is, a data set where every row and column represents a charity, and the cells indicate whether a pair of charities are linked through at least one trustee.

First, let's see how many charities are in the data:

In [95]:
len(data.drop_duplicates("regno")) # how many unique charity ids are there in the data?

1123

OK, that means we'll have a matrix with 1123 rows and 1123 columns (one for each charity). If a matrix has the same number of rows and columns it is known as a *square matrix*.

#### Adjacency matrix

Let's use some clever Python code to create an adjacency matrix where the nodes are charities - that is, every row represents a charity, as does every column.

In [96]:
data_merge = data.merge(data, on="trustee_id")
charity_mat = pd.crosstab(data_merge.regno_x, data_merge.regno_y)
np.fill_diagonal(charity_mat.values, 0)
charity_mat[charity_mat >= 1] = 1
charity_mat

regno_y  208879   209174   210037   210563   212479   212755   213258   \
regno_x                                                                  
208879         0        0        0        0        0        0        0   
209174         0        0        0        0        0        0        0   
210037         0        0        0        0        0        0        0   
210563         0        0        0        0        0        0        0   
212479         0        0        0        0        0        0        0   
...          ...      ...      ...      ...      ...      ...      ...   
1188334        0        0        0        0        0        0        0   
1188662        0        0        0        0        0        0        0   
1188791        0        0        0        0        0        0        0   
1188851        0        0        0        0        0        0        0   
1188892        0        0        0        0        0        0        0   

regno_y  214684   215728   216533   ...  1186986  1187340  1187446  1187493  \
regno_x                             ...                                       
208879         0        0        0  ...        0        0        0        0   
209174         0        0        0  ...        0        0        0        0   
210037         0        0        0  ...        0        0        0        0   
210563         0        0        0  ...        0        0        0        0   
212479         0        0        0  ...        0        0        0        0   
...          ...      ...      ...  ...      ...      ...      ...      ...   
1188334        0        0        0  ...        0        0        0        0   
1188662        0        0        0  ...        0        0        0        0   
1188791        0        0        0  ...        0        0        0        0   
1188851        0        0        0  ...        0        0        0        0   
1188892        0        0        0  ...        0        0        0        0   

regno_y  1188245  1188334  1188662  1188791  1188851  1188892  
regno_x                                                        
208879         0        0        0        0        0        0  
209174         0        0        0        0        0        0  
210037         0        0        0        0        0        0  
210563         0        0        0        0        0        0  
212479         0        0        0        0        0        0  
...          ...      ...      ...      ...      ...      ...  
1188334        0        0        0        0        0        0  
1188662        0        0        0        0        0        0  
1188791        0        0        0        0        0        0  
1188851        0        0        0        0        0        0  
1188892        0        0        0        0        0        0  

[1123 rows x 1123 columns]

As is probably obvious at the outset, there are lots of zeroes; that is, most charities are not connected to most other charities. But they are all connected to at least one other: we can see this by calculating the row/column totals:

In [84]:
charity_mat["row_total"] = charity_mat.sum(axis=1)
charity_mat["row_total"].describe()

count    1123.000000
mean        2.746215
std         2.679671
min         1.000000
25%         1.000000
50%         2.000000
75%         3.000000
max        23.000000
Name: row_total, dtype: float64

In [85]:
charity_mat["col_total"] = charity_mat.sum(axis=0)
charity_mat["col_total"].describe()

count    1123.000000
mean        2.746215
std         2.679671
min         1.000000
25%         1.000000
50%         2.000000
75%         3.000000
max        23.000000
Name: col_total, dtype: float64

[*The above totals are somewhat nonsensical - surely not every charity is linked to at least one other? Is it an issue with data collection?*]

#### Edge list

In [97]:
charity_el = charity_mat.stack().reset_index() # convert to edgelist
charity_el = charity_el[charity_el[0]==1] # keep pairs of charities that are connected
charity_el = charity_el.drop(0, axis=1)
charity_el.rename(columns = {"regno_x": "source", "regno_y": "target"}, inplace = True)

In [98]:
charity_el[charity_el["target"]==208879]

source  target
143744    278469  208879
157220    284334  208879
751287   1123968  208879
911876   1143295  208879
1038775  1160228  208879

In [99]:
charity_el[charity_el["source"]==208879]

source   target
128  208879   278469
140  208879   284334
669  208879  1123968
812  208879  1143295
925  208879  1160228

In [100]:
len(charity_el)

3084

We can see that each connection between a pair of charities is listed twice - this is because we are dealing with undirected ties: charity *208879* is linked to charity *278469*, therefore charity *278469* is also linked to charity *208879*. We should delete these duplicates before progressing any further:

In [101]:
edgelist = charity_el[pd.DataFrame(np.sort(charity_el[['source','target']].values,1)).duplicated().values]
len(edgelist)

1542

[*Need some way of checking the above code*]

In [103]:
edgelist[edgelist["source"]==278469]

source  target
143744  278469  208879

## A Simple Analysis

Let's bring some of our fundamental concepts to life with a simple research example. [Manchester trustees]



In [104]:
import networkx as nx

In [108]:
G = nx.from_pandas_edgelist(edgelist, "source", "target")
print(nx.info(G)) # get basic info about network

Name: 
Type: Graph
Number of nodes: 1123
Number of edges: 1542
Average degree:   2.7462


Good, we can see that the number of nodes and ties (edges) match the objects we created earlier:

In [109]:
print(len(charity_mat), len(edgelist))

1123 1542


## Conclusion

*SNA and its value and limitations and opportunities*.

Good luck on your data-driven travels!

## Bibliography

Barba, Lorena A. et al. (2019). *Teaching and Learning with Jupyter*. <a href="https://jupyter4edu.github.io/jupyter-edu-book/" target=_blank>https://jupyter4edu.github.io/jupyter-edu-book/</a>.

Brooker, P. (2020). *Programming with Python for Social Scientists*. London: SAGE Publications Ltd.

Lau, S., Gonzalez, J., & Nolan, D. (n.d.). *Principles and Techniques of Data Science*. https://www.textbook.ds100.org

Tagliaferri, L. (n.d.). *How to Code in Python 3*. https://assets.digitalocean.com/books/python/how-to-code-in-python.pdf

## Further reading and resources

We publish a list of useful books, papers, websites and other resources on our web-scraping Github repository: <a href="https://github.com/UKDataServiceOpen/social-network-analysis/tree/master/reading-list/" target=_blank>[Reading list]</a>

The help documentation for the `networkx` and `pandas` modules is refreshingly readable and useful:
* <a href="LINK" target=_blank>`networkx`</a>
* <a href="LINK" target=_blank>`pandas`</a> 

You may also be interested in the following articles and lessons specifically relating to social network analysis:
* <a href="https://programminghistorian.org/en/lessons/exploring-and-analyzing-network-data-with-python" target=_blank>Exploring and Analyzing Network Data with Python </a>
* <a href="https://programminghistorian.org/en/lessons/creating-network-diagrams-from-historical-sources" target=_blank>From Hermeneutics to Data to Networks: Data Extraction and Network Visualization of Historical Sources</a>

## Appendices